[PORT] Port of shielded component

Sarchutar opened the pull request:
## About The Pull Request
Ports:
- https://github.com/tgstation/tgstation/pull/57797
- https://github.com/tgstation/tgstation/pull/58757

Original description:

> Shielded hardsuits (like the syndie ones) and shielded cult robes, despite functioning very similarly, were actually implemented twice in their own pockets of clothing code. This merges them into one component that lets you block a certain number of attacks while your suit has charges, and have the suit recharge itself after going so long without being hit (optional, cult robes still don't regain lost charges).
> ....
> One casualty of this is that you can no longer toggle the shield color of syndie hardsuits with a multitool since that it was annoying to componentize, not something that affected gameplay, and probably something noone knew you could do anyway.


## Why It's Good For The Game
Makes it easier to add shields to things in the future

## Testing Photographs and Procedure
<details>
<summary>Screenshots&Videos</summary>

https://user-images.githubusercontent.com/53494785/184227818-6d7978c6-57c2-4b8e-aa81-aea67f6d0448.mp4

</details>

## Changelog
:cl: Sarchutar, Ryll-Ryll
add: Shields for hardsuits (Syndie, CTF, Wizard etc) are now a component
del: Shielded Syndicate Hardsuits cannot have their shield color changed with a multitool anymore
/:cl:

Sarchutar committed:
port of shielding component

- ports tg#57797 and tg#58757

PowerfulBacon commented on code/datums/components/shielded.dm:
```diff
@@ -0,0 +1,160 @@
+	. = COMPONENT_NO_AFTERATTACK
+	if(!istype(parent, /obj/item/clothing/suit/space/hardsuit/shielded/wizard))
+		to_chat(user, "<span class='warning'>The rune can only be used on battlemage armour!</span>")
+		return
```
This code should be generic, since its part of a generic component. There shouldn't be specific type checking.

Sarchutar committed:
review things

- doesn't restrict the recharging to only wizard armor charges should someone later decide to add more

PowerfulBacon commented on code/datums/components/shielded.dm:
```diff
@@ -0,0 +1,160 @@
+/datum/component/shielded/proc/check_recharge_item(datum/source, obj/item/item, mob/living/user)
+	SIGNAL_HANDLER
+
+	if(istype(item, /obj/item/wizard_armour_charge))
+		. = COMPONENT_NO_AFTERATTACK
+		var/obj/item/wizard_armour_charge/recharge_rune = item
+		if(!istype(parent, /obj/item/clothing/suit/space/hardsuit/shielded/wizard))
+			to_chat(user, "<span class='warning'>The rune can only be used on battlemage armour!</span>")
+			return
+
+		current_charges += recharge_rune.restored_charges
+		to_chat(user, "<span class='notice'>You charge \the [parent]. It can now absorb [current_charges] hits.</span>")
+		qdel(recharge_rune)
```
Perhaps something like /obj/item/armour_charge which has a variable list of the types of armour it affects, then creating a wizard amour charge for that would be cleaner, but I guess this works and I'm happy to merge with this method due to it being unlikely to be expanded.
